drv/lpc55-sprot-server/src/main.rs

```diff
@@ -220,132 +209,136 @@ impl Io {
             let intstat = self.spi.intstat();
             if intstat.ssa().bit() {
                 self.spi.ssa_clear();
-                break self.tight_loop(&mut tx_iter, rx_msg);
+                break;
             }
-        };
-
-        if signal_reply {
-            self.deassert_rot_irq();
         }
-
-        result
     }
 
-    // Read data in a tight loop until we see CSn de-asserted
-    //
-    // XXX Denial of service by forever asserting CSn?
-    // We could mitigate by imposing a time limit
-    // and resetting the SP if it is exceeded.
-    // But, the management plane is going to notice that
-    // the RoT is not available. So, does it matter?
-    fn tight_loop<'a>(
+    pub fn wait_for_request(
         &mut self,
-        tx_iter: &mut dyn Iterator<Item = &u8>,
-        rx_msg: &mut RxMsg<'a>,
-    ) -> Result<(), IoError> {
-        let mut err = None;
+        rx_buf: &mut [u8],
+    ) -> Result<usize, IoError> {
+        self.wait_for_csn_asserted(false);
+
+        // Go into a tight loop receiving as many bytes as we can until we see
+        // CSn de-asserted.
+        let mut bytes_received = 0;
+        let mut rx = rx_buf.iter_mut();
+        while !self.spi.ssd() {
+            while self.spi.has_byte() {
+                bytes_received += 1;
+                let read = self.spi.read_u8();
+                rx.next().map(|b| *b = read);
+            }
+        }
 
-        let mut csn_deasserted = false;
-        let mut too_many_bytes_received = false;
+        self.spi.ssd_clear();
 
-        loop {
-            // Let's exchange some bytes
-            let mut io = true;
-            while io {
-                io = false;
-                if self.spi.can_tx() {
-                    io = true;
-                    if let Some(b) = tx_iter.next().copied() {
-                        self.spi.send_u8(b);
-                    } else {
-                        // Just clock out zeros and prevent an unnecessary underrun
-                        self.spi.send_u8(0);
-                    }
-                }
-
-                if self.spi.has_byte() {
-                    io = true;
-                    let b = self.spi.read_u8();
-                    if rx_msg.push(b).is_err() {
-                        too_many_bytes_received = true;
-                    }
-                }
-            }
+        // There may be bytes left in the rx fifo after CSn is de-asserted
+        while self.spi.has_byte() {
+            bytes_received += 1;
+            let read = self.spi.read_u8();
+            rx.next().map(|b| *b = read);
+        }
 
-            // Let's check for any problems
-            let fifostat = self.spi.fifostat();
+        self.check_for_overrun()?;
```

Comment:
Nit: the term "overrun" is a bit ambiguous, referring to this as the hardware does (as RX Error) would make it easier to compare to the datasheet. Y'know, for the two of us who will ever do that. :-)

Reply:
Will do!